### ISSUE TITLE
Manually closing the qt viewer causes qt RuntimeErrors when re-adding layers

### ISSUE THREAD

**Issue by brisvag** [napari/napari#2110]
## 🐛 Bug

If the viewer is closed manually from the os side (but not when using `viewer.close()`), layers that were present in the viewer will be "tainted", and cause a few runtime errors if later re-added to another viewer.

## To Reproduce
Normal behaviour:
```python
pl = napari.layers.Points()

v = napari.Viewer()
v.layers.append(pl)
# close it programmatically
v.close()
# make a new viewer and append the layer
v = napari.Viewer()
v.layers.append(pl)
# click anywhere on the layerlist widget: nothing bad happens (foreshadowing...).
```
If closed manually, bugs arise:
```python
pl = napari.layers.Points()

v = napari.Viewer()
v.layers.append(pl)
# close it manually...
# then create a new one and append the layer
v = napari.Viewer()
v.layers.append(pl)
```
After the last step, one of two things will happen:
- if before closing the viewer you *deselected* the layer, this error is raised:
```python
.../lib/python3.9/site-packages/napari/_qt/qt_main_window.py in _status_changed(self, event)
    867             The napari event that triggered this method.
    868         """
--> 869         self._status_bar.showMessage(event.value)
    870 
    871     def _title_changed(self, event):

RuntimeError: wrapped C/C++ object of type QStatusBar has been deleted
```
- if you left the layer *selected*, nothing happens until you click anywhere on the layerlist widget, which fires the same error. This however appears in a weird way in ipython, a non-colored, "streaming" kind of way that I don't quite understand.

## Environment
```yaml
napari: 0.4.3rc2
Platform: Linux-5.10.6-arch1-1-x86_64-with-glibc2.32
Python: 3.9.1 (default, Dec 13 2020, 11:55:53) [GCC 10.2.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.19.5
SciPy: 1.6.0
Dask: 2.30.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 460.32.03
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen #1: resolution 1920x1080, scale 1.0

Plugins:
- napari_plugin_engine: 0.1.9
- svg: 0.1.4
```

**Comment by sofroniewn** [comment]
Hmm, I think if we figure out how to get the manual close to trigger the `viewer.close` call then this should work!

**Comment by tlambert03** [comment]
clicking the manual close button at the top left should call the [`closeEvent()`](https://doc.qt.io/qt-5/qwidget.html#closeEvent) method on the `QMainWindow`.  as a quick and dirty, should be able to hook into that to call `viewer.close()`, (but this also may be indicative of a slightly poor arrangement that has a "better" fix)

edit: specifically, we'd add manual method override [here](https://github.com/napari/napari/blob/87961d0554b2bb1574553e23bf2231a9a5117568/napari/_qt/qt_main_window.py#L46)

**Comment by sofroniewn** [comment]
@goanpeca there may be an opportunity now to cleanly fix this after the merge of #2397, do you have ideas - it sound like in zulip @brisvag is running into some infinite recursion problems with just triggering viewer.close directly :-)

https://napari.zulipchat.com/#narrow/stream/270578-reviews/topic/.232110/near/230009888

**Comment by sofroniewn** [comment]
@ianhi you might be interested in this issue too based on your comments here https://napari.zulipchat.com/#narrow/stream/212875-general/topic/on-close/near/230088585

**Comment by sofroniewn** [comment]
Just noting @tlambert03 thinks this will be fixed by #3184 when that is merged

**Comment by brisvag** [comment]
I don't know when this was fixed, byt it's fine now! Git bisect was not helpful, it seems to be fixed way back...
